### PR TITLE
Teach pegen to compute first sets

### PR DESCRIPTION
**Pull request by pablogsal**

### CODE REVIEW

**Comment by gvanrossum** [comment]
How are you planning to use this?

**Comment by pablogsal** [comment]
> How are you planning to use this?

Is a first step towards exploring what was discussed in https://github.com/gvanrossum/pegen/issues/49 regarding [this paper](http://kmizu.github.io/papers/paste513-mizushima.pdf). Ideally, we could also use this information for introducing lookaheads. Mainly I thought is nice to have to also compare with the original grammar

**Comment by gvanrossum** [comment]
Hm, #49 is mostly about something with 'cut' that I fail to understand, and I'm not sure I follow the paper either. But I do have an idea for using FIRST sets to optimize the parser, and this is a first step.

However, an important reason why FIRST set calculation is subtle is that it should properly handle rules with empty productions. I have this example:
```
start: sign thing $
sign: ['-']
thing: NUMBER
```
and your program claims that the FIRST set for `start` is just `{"'-'"}`, whereas it should include `NUMBER` as well. Also, the FIRST set for `sign` should indicate that it may be empty. I think traditionally there's a symbol "epsilon" that's used for this; we could use either an empty string or `None`.

(You could argue that there's something wrong with this grammar, because we don't expect empty productions, but I think it may be better to just deal with those -- it's slightly complicated to prove that a rule *doesn't* have an empty production.)

**Comment by gvanrossum** [comment]
FWIW, parser_generator.py has a function compute_nullables() that can help with this.

**Comment by apalala** [comment]
Please see [this pull request on TatSu](https://github.com/neogeny/TatSu/pull/139) for a full discussion, and a promising implementation.

I agree with @gvanrossum that his simplification of the nullability algorithm is enough to detect and resolve left-rec.

**Comment by gvanrossum** [comment]
@pablogsal Are you waiting for someone or something?

**Comment by pablogsal** [comment]
> @pablogsal Are you waiting for someone or something?

No, I was finishing other stuff. I will land this now.

File: pegen/first_sets.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+            # If the set of new terminals can start with the empty string,
+            # it means that the item is completelly nullable and we should
+            # also considering at least the next item in case the current
+            # one fails to parse.
+
+            if "" in new_terminals:
+                continue
+
+            if not isinstance(other.item, (Opt, NegativeLookahead, Repeat0)):
```
Comment (apalala) [inline pegen/first_sets.py]: This PR has already been merged, but `Cut` seems to be missing from this list.

File: pegen/first_sets.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            result |= new_terminals
+            if to_remove:
+                result -= to_remove
+
+            # If the set of new terminals can start with the empty string,
+            # it means that the item is completelly nullable and we should
+            # also considering at least the next item in case the current
+            # one fails to parse.
+
+            if "" in new_terminals:
```
Comment (apalala) [inline pegen/first_sets.py]: Perhaps:
```python
if "" not in new_terminals:
    break
```

**Pull request merged**